instrumentation/nats/nats-2.21/library/src/test/java/io/opentelemetry/instrumentation/nats/v2_21/NatsTelemetryDispatcherTest.java

```diff
@@ -0,0 +1,155 @@
+        trace -> assertReceiveProcessSpans(trace, "sub2"),
+        trace -> assertReceiveProcessSpans(trace, "sub3"));
+
+    assertThatNoException()
+        .isThrownBy(
+            () -> {
+              d1.unsubscribe("sub1");
+              d1.unsubscribe("sub2", 1);
+            });
+  }
+
+  @Test
+  void testSubscribeCustomHandler() {
+    // given
+    TestConnection testConnection = new TestConnection();
+    OpenTelemetryConnection connection = telemetry.wrap(testConnection);
+    Dispatcher dispatcher = connection.createDispatcher();
+    Subscription s1 = dispatcher.subscribe("sub1", handler);
+    Subscription s2 = dispatcher.subscribe("sub2", "queue", handler);
+    Subscription s3 = connection.createDispatcher(handler).subscribe("sub3", handler);
+
+    // when
+    testConnection.deliver(NatsMessage.builder().subject("sub1").data("x").build());
+    testConnection.deliver(NatsMessage.builder().subject("sub2").data("x").build());
+    testConnection.deliver(NatsMessage.builder().subject("sub3").data("x").build());
+
+    // then
+    testing.waitAndAssertTraces(
+        trace -> assertReceiveProcessSpans(trace, "sub1"),
+        trace -> assertReceiveProcessSpans(trace, "sub2"),
+        trace -> assertReceiveProcessSpans(trace, "sub3"));
+
+    // and
+    assertThatNoException()
+        .isThrownBy(
+            () -> {
+              s1.unsubscribe();
+              dispatcher.unsubscribe(s2);
+              s3.unsubscribe(1);
+            });
+  }
+
+  @Test
+  void testSpanLink() {
+    // given
+    TestConnection testConnection = new TestConnection();
+    OpenTelemetryConnection connection = telemetry.wrap(testConnection);
+    connection.createDispatcher(handler).subscribe("sub");
+
+    String linkTraceId = "0af7651916cd43dd8448eb211c80319c";
+    Headers headers =
+        new Headers(
+            new Headers().put("traceparent", "00-" + linkTraceId + "-b7ad6b7169203331-01"), true);
+
+    // when
+    testConnection.deliver(NatsMessage.builder().subject("sub").data("x").headers(headers).build());
+
+    // then
+    testing.waitAndAssertTraces(
+        trace ->
+            trace.hasSpansSatisfyingExactly(
+                span ->
+                    span.hasName("sub receive")
+                        .hasNoParent()
+                        .hasLinksSatisfying(
+                            links ->
+                                assertThat(links)
+                                    .singleElement()
+                                    .satisfies(
+                                        link ->
+                                            assertThat(link.getSpanContext().getTraceId())
+                                                .isEqualTo(linkTraceId))),
+                span -> span.hasName("sub process").hasParent(trace.getSpan(0))));
+  }
+
+  private static void assertReceiveProcessSpans(TraceAssert trace, String subject) {
+    trace.hasSpansSatisfyingExactly(
```

Comment:
why no publish span?
At first glance I'd say the trace structure does not conform to what our other messaging instrumentations produce. By default we don't create receive spans and span links, these are enabled with `-Dotel.instrumentation.messaging.experimental.receive-telemetry.enabled=true` flag. Is there a reason why you wish to deviate from what the other instrumentations do?

Reply:
I have no intention of deviating, no, sorry for this. I wrongly thought that the experimental was the end goal of the messaging instrumentation and went straight away for the "final" implementation. 

Dispatcher is the consuming side of NATS. You create a dispatcher and you subscribe it to one or more subjects(=topics if we do a parallel with Kafka. Dispatcher would be a consumer). 

So if I get this right, I just need to implement the flag properly so it creates another trace with a receive span and a span link. And by default, we'll just get the process spans in the same trace, right?